WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/FIle List/ProductDownloadListViewModel.swift

```diff
@@ -0,0 +1,123 @@
+}
+
+/// Error cases that could occur in product downloadable files settings.
+///
+enum ProductDownloadListError: Error {
+    case emptyFileName
+    case emptyUrl
+}
+
+/// Provides view data for downloadable files settings, and handles init/UI/navigation actions needed in product downloadable files settings.
+///
+final class ProductDownloadListViewModel: ProductDownloadListViewModelOutput {
+    private let product: ProductFormDataModel
+
+    // Editable data
+    //
+    private(set) var downloadableFiles = [ProductDownloadDragAndDrop]()
+    private(set) var downloadLimit: Int64
+    private(set) var downloadExpiry: Int64
+
+    init(product: ProductFormDataModel) {
+        self.product = product
+
+        downloadableFiles = product.downloadableFiles.map { ProductDownloadDragAndDrop(downloadableFile: $0) }
+        downloadLimit = product.downloadLimit
+        downloadExpiry = product.downloadExpiry
+    }
+
+    // MARK: - ProductDownloadListDataSource Methodes
+    //
+    func remove(at index: Int) -> ProductDownloadDragAndDrop? {
+        return downloadableFiles.remove(at: index)
+    }
+
+    func item(at index: Int) -> ProductDownloadDragAndDrop? {
+        return downloadableFiles[index]
+    }
+
+    func insert(_ newElement: ProductDownloadDragAndDrop, at index: Int) {
+        downloadableFiles.insert(newElement, at: index)
+    }
+
+    func count() -> Int {
+        return downloadableFiles.count
+    }
+}
+
+extension ProductDownloadListViewModel: ProductDownloadListActionHandler {
+
+    // MARK: - Tap actions
+
+    func didTapDownloadableFileFromRow(_ indexPath: IndexPath) {
+        //TODO: Show respective file in a new window
+    }
+
+    // MARK: - UI changes
+
+    // Input field actions
+    func handleDownloadsChange(_ downloads: [ProductDownload]) {
+        self.downloadableFiles = downloads.map { ProductDownloadDragAndDrop(downloadableFile: $0) }
+    }
+
+    func handleDownloadLimitChange(_ downloadLimit: Int64) {
+        self.downloadLimit = downloadLimit
+    }
+
+    func handleDownloadExpiryChange(_ downloadExpiry: Int64) {
+        self.downloadExpiry = downloadExpiry
+    }
+
+    // Navigation actions
+    func completeUpdating(onCompletion: ProductDownloadListViewController.Completion) {
+        //TODO: Perform data validation as necessary
+        let data = ProductDownloadsEditableData(downloadableFiles: downloadableFiles.map { $0.downloadableFile },
+                                                downloadLimit: downloadLimit,
+                                                downloadExpiry: downloadExpiry)
+        onCompletion(data)
+    }
+
+    func hasUnsavedChanges() -> Bool {
+        if downloadLimit != product.downloadLimit ||
+            downloadableFiles.count != product.downloadableFiles.count ||
+            downloadExpiry != product.downloadExpiry {
+            return true
+        }
+
+        //TODO: Check if the data has been changed and return accordingly.
```

Comment:
This should be implemented in this PR, or we will forget about it. You should just compare the `downloadableFiles` array 👍

Reply:
Great. Updated the logic